SimplifiedTests/NYPLReaderBookmarksBusinessLogicTests.swift

```diff
@@ -194,8 +198,206 @@ class NYPLReaderBookmarksBusinessLogicTests: XCTestCase {
       }
     }
 
-    // MARK: Helper
+  // MARK: - Test addBookmark/postBookmark
+
+  func testAddBookmarkWithSucceededUpload() throws {
+    // Make sure server contains no bookmark
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      XCTAssertEqual(bookmarks, nil)
+    }
+    
+    let bookmarkLoc = newBookmarkR2Location(href: "Intro", chapter: "1", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    guard let _ = bookmarkBusinessLogic.addBookmark(bookmarkLoc) else {
+      XCTFail("Failed to create new bookmark")
+      return
+    }
+    
+    // There should be one bookmark uploaded
+    XCTAssertEqual(self.bookRegistryMock.readiumBookmarks(forIdentifier: self.bookIdentifier).count, 1)
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      guard let bookmarks = bookmarks else {
+        XCTFail("Failed to get bookmark from server")
+        return
+      }
+      XCTAssertEqual(bookmarks.count, 1)
+    }
+  }
+
+  func testAddBookmarkWithFailedUpload() throws {
+    // Make sure server contains no bookmark
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      XCTAssertEqual(bookmarks, nil)
+    }
+    
+    let bookmarkLoc = newBookmarkR2Location(href: "Intro", chapter: "1", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    annotationsMock.failRequest = true
+    guard let _ = bookmarkBusinessLogic.addBookmark(bookmarkLoc) else {
+      XCTFail("Failed to create new bookmark")
+      return
+    }
+    
+    // BookRegistry should have one bookmark even upload failed
+    // While server should not have any bookmarks
+    XCTAssertEqual(self.bookRegistryMock.readiumBookmarks(forIdentifier: self.bookIdentifier).count, 1)
+    
+    annotationsMock.failRequest = false
```

Comment:
do we need to set this flag again, since we set it a few lines before?

Reply:
We set it to true above so the bookmark post request will fail, we need to set it back to false in order to get the number of bookmarks from the mocked server.